### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by gohabereg**

### CODE REVIEW

File: src/Capella.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+namespace Capella;
+
+
+/**
+ * Class Uploader
+ * @package Capella
+ *
+ * PHP SDK for CodeX Capella
+ *
+ * @example
```
Comment (neSpecc) [inline src/Capella.php]: нужен readme и подключение через composer

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/Capella.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+     * Send cURL request to capella service
+     *
+     * @param string $url
+     * @param array $params
+     * @param string $method
+     * @return mixed
+     * @throws \Exception
+     *
+     */
+    protected function sendRequest($url, array $params, $method='POST') {
```
Comment (neSpecc) [inline src/Capella.php]: может private?

File: src/Capella.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+                break;
+
+            case 'GET':
+                if (!count($params)) {
+                    break;
+                }
+
+                $query = '';
+                foreach ($params as $key => $value) {
```
Comment (neSpecc) [inline src/Capella.php]: http://php.net/manual/en/function.http-build-query.php

File: src/Capella.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
+
+        $result = curl_exec($curl);
+        curl_close($curl);
+
+        return $result;
+
+    }
+
+    protected function loadConfig() {
```
Comment (neSpecc) [inline src/Capella.php]: private

File: src/Capella.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace Capella;
+
+
+/**
+ * Class Uploader
+ * @package Capella
```
Comment (neSpecc) [inline src/Capella.php]: надо задокументировать авторство codex team и ссылку на гитхаб

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/Capella.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+     * Upload image to capella service
+     *
+     * @param string $path - path or url to image
+     * @return mixed
+     */
+    public function upload($path) {
+
+        $url = $this->API_URL;
+
+        if (file_exists($path)) {
```
Comment (n0str) [inline src/Capella.php]: If the file does not exist (for ex. due to misprint) system decides that it's a URL. But it's not an URL.

**Review by n0str — CHANGES_REQUESTED** [review]

**Review by talyguryn — APPROVED** [review]

**Review by n0str — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+## Usage
+
+First, read CodeX Capella [documentation](https://github.com/codex-team/capella#readme).
+
+#### Uploading
+```php
+use \Capella\Capella;
+
+$image = Capella::upload('path-or-url-to-your-pipcture');
```
Comment (neSpecc) [inline README.md]: picture

File: README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+$url = $image->url();
+```
+
+#### Get image by known id
+```php
+use \Capella\Capella;
+
+$image = Capella::image('abcdef-1234-abcd-1234');
```
Comment (neSpecc) [inline README.md:32]: Provide link to the `static image($id)` SDK method description where User can read about parameters return value

File: README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+First, read CodeX Capella [documentation](https://github.com/codex-team/capella#readme).
+
+#### Uploading
+```php
+use \Capella\Capella;
+
+$image = Capella::upload('path-or-url-to-your-pipcture');
+
+$url = $image->url();
```
Comment (neSpecc) [inline README.md]: Provide link to the `static upload()` SDK method description where User can read about parameters return value

File: README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+$image = Capella::upload('picture.jpg');
+
+$url = $image->resize(100)->url();
+echo $url; // https://capella.pics/<id>/resize/100
+  
+$url = $image->crop(100, 200);
+echo $url; // https://capella.pics/<id>/crop/100x200
+```
+
```
Comment (neSpecc) [inline README.md:50]: Add link to the SDK methods description

File: docs/sdk.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+## Classes
+
+ - [`\Capella\Capella`](#capellacapella)
+ - [`\Capella\Uploader`](#capellauploader)
+ - [`\Capella\CapellaImage`](#capellacapellaimage)
+ - [`\Capella\CapellaException`](#capellacapellaexception)
+
+## `\Capella\Capella`
+
+This class provides two basic methods to work with Capella API.
```
Comment (neSpecc) [inline docs/sdk.md]: methods for working with

File: docs/sdk.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+ 
+ #### `static upload($path)`
+ 
+ Upload image file to capella.pics, or send image URL.
```
Comment (neSpecc) [inline docs/sdk.md]: Upload an `Image file` or send an `Image URL` to the Capella

File: docs/sdk.md (hunk 11 above)
Comment (neSpecc) [reply docs/sdk.md]: `[Capella API](https://github.com/codex-team/capella#readme)`

File: docs/sdk.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+ 
+| Parameter | Type | Description |
+| --------- | ---- | ----------- |
+| `path`    | `string` | Image URL or path to upload to capella.pics |
```
Comment (neSpecc) [inline docs/sdk.md]: Image URL of File path that should be uploaded to the Capella

File: docs/sdk.md [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+
+Return [`CapellaImage`](#capellacapellaimage) object if upload was successful.
+Otherwise throws [`CapellaImage`](#capellacapellaexception).
```
Comment (neSpecc) [inline docs/sdk.md]: the CapellaException

File: docs/sdk.md [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+| --------- | ---- | ----------- |
+| `path`    | `string` | Image URL or path to upload to capella.pics |
+
+
+Return [`CapellaImage`](#capellacapellaimage) object if upload was successful.
+Otherwise throws [`CapellaImage`](#capellacapellaexception).  
+
+#### `static image($id)`
+
+Get [`CapellaImage`](#capellacapellaimage) object with passed `$id`
```
Comment (neSpecc) [inline docs/sdk.md]: by passed

File: docs/sdk.md [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+| Parameter | Type     | Description |
+| --------- | -------- | ----------- |
+| `id`      | `string` | Image id, received from capella.pics |
```
Comment (neSpecc) [inline docs/sdk.md]: unnecessary comma

File: docs/sdk.md [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+Return [`CapellaImage`](#capellacapellaimage) object.
```
Comment (neSpecc) [inline docs/sdk.md]: the

File: docs/sdk.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+## `\Capella\Uploader`
+
+Upload images to capella using cURL.
```
Comment (neSpecc) [inline docs/sdk.md]: to the Capella

File: docs/sdk.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+#### `upload($path)`
+| Parameter | Type     | Description |
+| --------- | -------- | ----------- |
+| `path`    | `string` | Image URL or path to upload to capella.pics |
+
+
+Return `stdClass` object if path is valid and capella.pics is accessible.
```
Comment (neSpecc) [inline docs/sdk.md]: and Capella

File: docs/sdk.md [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+| Field     | Description |
+| --------- | ----------- |
+| `success` | Upload status |
+| `message` | Error or success message |
+| `id`      | Uploaded image id |
+| `url`     | Uploaded image url |
```
Comment (neSpecc) [inline docs/sdk.md]: URL

File: docs/sdk.md [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+
+Throws [`CapellaImage`](#capellacapellaexception) if `path` is not existing file or valid and accessible URL.
```
Comment (neSpecc) [inline docs/sdk.md]: the CapellaImageException (or CapellaException )

Comment (neSpecc) [reply docs/sdk.md]: > path is not existing file or valid and accessible URL. 

if passed `path` either invalid file link or inaccessible URL

File: docs/sdk.md [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+Throws [`CapellaImage`](#capellacapellaexception) if `path` is not existing file or valid and accessible URL. 
+
+## `\Capella\CapellaImage`
+
+#### Properties
+
+| Property | Description |
+| -------- | ----------- |
+| `id`     | Image id    |
+| `url`    | Base image url |
```
Comment (neSpecc) [inline docs/sdk.md]: URL

File: docs/sdk.md [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+
+
+#### `__constructor($id)`
+
+Set class properties
+
+| Parameter | Type     | Description |
+| --------- | -------- | ----------- |
+| `id`      | `string` | Image id, received from capella.pics |
```
Comment (neSpecc) [inline docs/sdk.md]: from Capella

File: docs/sdk.md [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+
+#### `clear()`
+
+Clear filters sequence.
+
+Return `CapellaImage` object with cleared filters
+
+#### `url()`
+
+Return url to image with applied filters.
```
Comment (neSpecc) [inline docs/sdk.md]: URL

Comment (neSpecc) [reply docs/sdk.md]: to the Image

File: README.md [hunk 25]
```diff
@@ -0,0 +1,70 @@
+# PHP SDK for CodeX Capella
+
+This repository provides PHP SDK for working with capella.pics API
```
Comment (neSpecc) [inline README.md]: with the Capella API

File: docs/sdk.md (hunk 16 above)
Comment (neSpecc) [reply docs/sdk.md]: from the Capella

File: src/Capella.php [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+require 'Uploader.php';
+require 'CapellaImage.php';
+require 'CapellaException.php';
+
+/**
+ * Class Capella
+ * @package Capella
+ *
+ * Base class for working with Capella API.
+ *
```
Comment (neSpecc) [inline src/Capella.php:15]: `@link <api-docs-link>`

File: src/Capella.php [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+ */
+class Capella
+{
+    private static $uploader = null;
+
+    private function __construct() {}
+
+    /**
+     * Upload image to capella.pics
```
Comment (neSpecc) [inline src/Capella.php]: to the Capella

File: src/Capella.php [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+     *
+     * Return CapellaImage class to apply filters
+     *
+     * @param string $path - path to image. Can be path to local file or external url
```
Comment (neSpecc) [inline src/Capella.php]: path to the Image. Can also be the Link to the local file or external URL

File: src/Capella.php [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+    public static function upload($path)
+    {
+        if (is_null(self::$uploader)) {
+            self::$uploader = new Uploader();
+        }
+
+        $response = self::$uploader->upload($path);
+
+        if (!$response || !$response->success) {
+            $errormsg = 'Upload to capella.pics failed.';
```
Comment (neSpecc) [inline src/Capella.php]: to the Capella

File: src/CapellaImage.php [hunk 30]
```diff
@@ -0,0 +1,171 @@
+<?php
+
+namespace Capella;
+
+/**
+ * Class CapellaImage
+ * @package Capella
+ *
+ * Class to work with image filters.
```
Comment (neSpecc) [inline src/CapellaImage.php]: for working

File: src/Uploader.php [hunk 31]
```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace Capella;
+
+
+/**
+ * Class Uploader
+ * @package Capella
+ *
+ * Provide image uploading to Capella service
```
Comment (neSpecc) [inline src/Uploader.php]: Image uploading to the

File: src/Uploader.php [hunk 32]
```diff
@@ -0,0 +1,159 @@
...
+     */
+    public function __construct()
+    {
+        $config = $this->loadConfig();
+
+        $this->API_URL = $config['api_url'] . 'upload';
+    }
+
+    /**
+     * Upload image to capella service
```
Comment (neSpecc) [inline src/Uploader.php]: to the Capella

File: src/Uploader.php [hunk 33]
```diff
@@ -0,0 +1,159 @@
...
+     *
+     * @param string $path - path or url to image
+     * @return \stdClass – array if path is valid and capella.pics is accessible or false
```
Comment (neSpecc) [inline src/Uploader.php]: Capella

Comment (neSpecc) [reply src/Uploader.php]: > or false

otherwise is false

File: src/Uploader.php [hunk 34]
```diff
@@ -0,0 +1,159 @@
...
+            return @json_decode($this->sendRequest($url, ['file' => $fileContext]));
+
+        } elseif ($this->isUrlValid($path)) {
+            /** Path is url */
+
+            return @json_decode($this->sendRequest($url, ['link' => $path]));
+
+        }
+
+        throw new CapellaException('Path is not existing file or valid and accessible URL.');
```
Comment (neSpecc) [inline src/Uploader.php]: Provided path links to the nonexistent/invalid file or inaccessible URL

File: src/Uploader.php [hunk 35]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    /**
+     *
+     * Send cURL request to capella service
```
Comment (neSpecc) [inline src/Uploader.php]: to the Capella

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: docs/sdk.md (hunk 15 above)
Comment (gohabereg) [reply docs/sdk.md]: Method doesn't get `CapellaImage` _by_ `id`. It just creates new `CapellaImage` object and pass `id` to the constructor.

**Review by neSpecc — APPROVED** [review]

**Pull request merged**